probe/endpoint/resolver.go

```diff
@@ -44,41 +63,64 @@ func NewReverseResolver() *ReverseResolver {
 // gcache.NotFoundKeyError error otherwise. Note: it returns one of the
 // possible names that can be obtained for that IP.
 func (r *ReverseResolver) Get(address string) (string, error) {
-	val, err := r.cache.Get(address)
-	if hostname, ok := val.(string); err == nil && ok {
-		return hostname, nil
-	}
-	if _, ok := val.(struct{}); err == nil && ok {
-		return "", errNotFound
-	}
-	if err == gcache.NotFoundKeyError {
+	val, found := r.cache.Get(address)
+	if !found {
 		// We trigger a asynchronous reverse resolution when not cached
 		select {
 		case r.addresses <- address:
 		default:
 		}
+		return "", errNotFound
+	}
+	if _, ok := val.(struct{}); ok {
+		return "", errNotFound
 	}
-	return "", errNotFound
+	entry, ok := val.(*resolverCacheEntry)
+	if !ok {
+		panic(fmt.Sprintf("unknown entry type in cache: %+v", val))
+	}
+	return entry.name, nil
 }
 
 func (r *ReverseResolver) loop() {
-	for request := range r.addresses {
-		// check if the answer is already in the cache
-		if _, err := r.cache.Get(request); err == nil {
-			continue
-		}
-		<-r.Throttle // rate limit our DNS resolutions
-		names, err := r.Resolver(request)
-		if err == nil && len(names) > 0 {
-			name := strings.TrimRight(names[0], ".")
-			r.cache.Set(request, name)
-		} else {
-			r.cache.Set(request, struct{}{})
+	purge := time.Tick(rAddrCacheExpiration)
+
+	for {
+		select {
+		case request := <-r.addresses:
+			// check if the answer is already in the cache
+			if _, ok := r.cache.Get(request); ok {
+				continue
+			}
+			<-r.Throttle // rate limit our DNS resolutions
+			names, err := r.Resolver(request)
+			if err == nil && len(names) > 0 {
+				name := strings.TrimRight(names[0], ".")
+				r.cache.Add(request, newResolverCacheEntry(name))
+			} else {
+				r.cache.Add(request, struct{}{})
+			}
+
+		case <-purge:
```

Comment:
I'm not asking for a new caching library.  I'm asking for a wrapper around the caching library we use so that we can share the timeout logic.